api/src/main/java/io/opentelemetry/context/Context.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.context;
+
+import io.opentelemetry.correlationcontext.CorrelationContext;
+import io.opentelemetry.trace.Span;
+
+// TODO (trask) once the current() method is removed (see below)
+//      this could be turned into an interface
+//      which improves interop possibilities with other libraries' contexts
+//      but also limits the ability to evolve this class with additional methods,
+//      e.g. withLoggingContext??? (insert sad Java 7 face here)
+//
+// TODO (trask) does this also need to support arbitrary key-values?
+//
+// TODO (trask) javadoc class and methods
+public abstract class Context {
+
+  public static Context empty() {
+    return DefaultContext.EMPTY;
+  }
+
+  // TODO (trask) remove this method and use CurrentContext.get() instead
+  //      so that this is a pure Context object, and has nothing to do with thread binding.
+  //      this changes was not done yet in order to reduce code churn
+  //      and make the PR easier to review as there are a lot of places
+  //      that call Context.current().
+  public static Context current() {
+    return DefaultContextStorage.current();
+  }
+
+  /**
+   * Creates a new {@link Context} with the given {@link Span} set.
+   *
+   * @param span the {@link Span} to be set.
+   * @return a new context with the given {@link Span} set.
+   * @since 0.5.0
+   */
+  public abstract Context withSpan(Span span);
```

Comment:
This breaks the layering principle to me, as well as creates a circular dependency between context layer and higher telemetry layers.

Reply:
Concur with this. Also, the idea is that when adding new concerns/layers, they all have their own `Context` related handling code (such as `TracingContextUtils`, which we can rename to something better, btw).

Reply:
It seems to create a more user-friendly API if we invert the dependency, and make the context depend on the "stuff", instead of making the "stuff" depend on the context:

E.g. here is some code in master:

```
    Context context =
        TracingContextUtils.withSpan(
            DefaultSpan.create(contextShim.getSpanContext()), Context.current());
    context =
        CorrelationsContextUtils.withCorrelationContext(
            contextShim.getCorrelationContext(), context);
```

and this is what is looks like in this PR:

```
    Context context =
        CurrentContext.get()
            .withSpan(DefaultSpan.create(contextShim.getSpanContext()))
            .withCorrelationContext(contextShim.getCorrelationContext());
```

Reply:
One could add additional an "user friendly context" layer that depends on both the low level context and "stuff" to provide a more friendly API.